src/psyclone/psyir/transformations/block_loop_trans.py

```diff
@@ -0,0 +1,237 @@
+
+        :returns: Tuple of None and None
+        :rtype: (None, None)
+        '''
+
+        self.validate(node, options)
+        if options is None:
+            options = {}
+        block_size = options.get("blocksize", 32)
+        # Create (or find) the symbols we need for the blocking transformation
+        routine = node.ancestor(nodes.Routine)
+        end_inner_loop = routine.symbol_table.symbol_from_tag(
+                "{0}_el_inner".format(node.variable.name),
+                symbol_type=DataSymbol,
+                datatype=node.variable.datatype)
+        outer_loop_variable = routine.symbol_table.symbol_from_tag(
+                "{0}_out_var".format(node.variable.name),
+                symbol_type=DataSymbol,
+                datatype=node.variable.datatype)
+        # We currently don't allow BlockedLoops to be ancestors of BlockedLoop
+        # so our ancestors cannot use these variables.
+
+        # Store the node's parent for replacing later and the start and end
+        # indicies
+        child0 = node.children[0]
+        child1 = node.children[1]
+
+        # For positive steps we do el_inner = min(out_var+block_size, el_outer)
+        # For negative steps we do el_inner = max(out_var-block_size, el_outer)
+        if int(node.children[2].value) > 0:
+            add = BinaryOperation.create(BinaryOperation.Operator.ADD,
+                                         Reference(outer_loop_variable),
+                                         Literal("{0}".format(block_size),
+                                                 node.variable.datatype))
+            minop = BinaryOperation.create(BinaryOperation.Operator.MIN, add,
+                                           child1.copy())
+            inner_loop_end = Assignment.create(Reference(end_inner_loop),
+                                               minop)
+        elif int(node.children[2].value) < 0:
+            sub = BinaryOperation.create(BinaryOperation.Operator.SUB,
+                                         Reference(outer_loop_variable),
+                                         Literal("{0}".format(block_size),
+                                                 node.variable.datatype))
+            maxop = BinaryOperation.create(BinaryOperation.Operator.MAX, sub,
+                                           child1.copy())
+            inner_loop_end = Assignment.create(Reference(end_inner_loop),
+                                               maxop)
+            # block_size needs to be negative if we're reducing
+            block_size = -block_size
+        # step size of 0 is caught by the validate call
+
+        # Replace the inner loop start and end with the blocking ones
+        child0.replace_with(Reference(outer_loop_variable))
+        child1.replace_with(Reference(end_inner_loop))
+
+        # Create the outerloop of the same type and loop_type
+        outerloop = Loop(variable=outer_loop_variable,
```

Comment:
Maybe we discussed last week and I forgot, but is there any reason this can not be `outerloop = Loop(variable=outer_loop_variable)` and then don't attempt to set the loop_type (since we already use the annotation to mark that it is a blocked loop). If that's valid, then you can remove the introduced valid_loop_types property.

Reply:
Perhaps I'd misunderstood, I though we decided to create this new loop's type to be the same type as the loop that is being transformed - the idea being to enable the transformation to be able to be API agnostic (whilst still being supported).

What I wanted to do was do `outerloop = type(node).create(...)`, however the `.create` function is not at all standardised, so this was my workaround. Its not perfect, as I'm not sure how psy.gen will work for these loop types if they're supported by APIs in someway, however my expectation is that they will probably work.

Reply:
Yeah that's fine then, I just wasn't sure if we say to reuse the same loop_type before or after we said the annotations could be used to mark these loops. I am not too concerned about this as probably will be revisited after the loop standarisation to make sure everything matches.